packages/beacon-node/src/sync/unknownBlock.ts

```diff
@@ -825,4 +753,231 @@ export class UnknownBlockSync {
+    this.peersMeta.delete(peerId);
+    this.activeRequests.delete(peerId);
+  }
+
+  /**
+   * called from fetchUnknownBlockRoot() where we only have block root and nothing else
+   * excludedPeers are the peers that we requested already so we don't want to try again
+   * pendingColumns is empty for prefulu, or the 1st time we we download a block by root
+   */
+  bestPeerForPendingColumns(pendingColumns: Set<number> | null, excludedPeers: Set<PeerIdStr>): PeerSyncMeta | null {
+    const eligiblePeers = this.filterPeers(pendingColumns, excludedPeers);
+    if (eligiblePeers.length === 0) {
+      return null;
+    }
+
+    const sortedEligiblePeers = sortBy(
+      shuffle(eligiblePeers),
+      // prefer peers with least active req
+      (peerId) => this.activeRequests.get(peerId) ?? 0
+    );
+
+    const bestPeerId = sortedEligiblePeers[0];
+    this.onRequest(bestPeerId);
+    return this.peersMeta.get(bestPeerId) ?? null;
+  }
+
+  /**
+   * called from fetchUnavailableBlockInput() where we have either BlockInput or NullBlockInput
+   * excludedPeers are the peers that we requested already so we don't want to try again
+   */
+  bestPeerForBlockInput(
+    unavailableBlockInput: BlockInput | NullBlockInput,
+    excludedPeers: Set<PeerIdStr>
+  ): PeerSyncMeta | null {
+    let cachedData: CachedData | undefined = undefined;
+    if (unavailableBlockInput.block === null) {
+      // NullBlockInput
+      cachedData = unavailableBlockInput.cachedData;
+    } else {
+      // BlockInput
+      if (unavailableBlockInput.type !== BlockInputType.dataPromise) {
+        throw Error(
+          `bestPeerForBlockInput called with BlockInput type ${unavailableBlockInput.type}, expected dataPromise`
+        );
+      }
+      cachedData = unavailableBlockInput.cachedData;
+    }
+
+    const eligiblePeers: PeerIdStr[] = [];
+
+    if (cachedData.fork === ForkName.fulu) {
+      // cached data is CachedDataColumns
+      const {dataColumnsCache} = cachedData;
+      const pendingDataColumns: Set<number> = new Set();
+      for (const column of this.custodyConfig.sampledColumns) {
+        if (!dataColumnsCache.has(column)) {
+          pendingDataColumns.add(column);
+        }
+      }
+      if (pendingDataColumns.size === 0) {
+        // no pending columns, we can return null
+        return null;
+      }
+      eligiblePeers.push(...this.filterPeers(pendingDataColumns, excludedPeers));
+    } else {
+      // prefulu
+      const pendingDataColumns = null;
+      eligiblePeers.push(...this.filterPeers(pendingDataColumns, excludedPeers));
+    }
+
+    if (eligiblePeers.length === 0) {
+      return null;
+    }
+
+    const sortedEligiblePeers = sortBy(
+      shuffle(eligiblePeers),
+      // prefer peers with least active req
+      (peerId) => this.activeRequests.get(peerId) ?? 0
+    );
+
+    const bestPeerId = sortedEligiblePeers[0];
+    this.onRequest(bestPeerId);
+    return this.peersMeta.get(bestPeerId) ?? null;
```

Comment:
This logic is duplicated from `bestPeerForPendingColumns`.  Do you think we should put this in a helper method?

Reply:
nah lets just keep it for now